[FLINK-16439][k8s] Make KubernetesResourceManager starts workers using WorkerResourceSpec requested by SlotManager

xintongsong opened the pull request:
## What is the purpose of the change
This PR is one step of FLINK-14106, making `KubernetesResourceManager` starts workers using `WorkerResourceSpec` requested by `SlotManager`.

This also means `KubernetesResourceManager` no longer:
- be aware of the default task executor resources
- assumes all workers are identical

This PR is based on #11320.

## Brief change log

- 8fcadbff7c80832626b317dc8cf75bc49b2a47d7..df11b25dc9094509f24cec5484a6c5b01622d3f8: Commits of previous PR.
- 76786fd6aa7e806f669eb3375ca8a3bd0534a0ab: Minor code clean-ups.
- d3fc9010a987cd299f355ddc1fbeb4225b5fb418: Introduce `PendingWorkerCounter` for counting pending workers per `WorkerResourceSpec` in `ActiveResourceManager`.
- d7a0626255f2445cfe20dcb7bab1aede3af4b0d7: KubernetesResourceManager starts workers with resources requested by SlotManager.

## Verifying this change

- Add `KubernetesResourceManagerTest#testStartAndRecoverVariousResourceSpec`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

xintongsong commented:
Updated PR extracting common logics into `ActiveResourceManager`, that could be reused by the Yarn setup.

KarmaGYZ commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ActiveResourceManager.java:
```diff
@@ -114,4 +121,42 @@ public ActiveResourceManager(
+			return pendingWorkerNums.getOrDefault(Preconditions.checkNotNull(workerResourceSpec), 0);
+		}
+
+		public int increateAndGet(final WorkerResourceSpec workerResourceSpec) {
```
increate -> increase.

KarmaGYZ commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -232,44 +229,57 @@ private void recoverWorkerNodesFromPreviousAttempts() throws ResourceManagerExce
+			createKubernetesTaskManagerParameters(workerResourceSpec);
+
+		podWorkerResources.put(parameters.getPodName(), workerResourceSpec);
+		final int pendingWorkerNum = pendingWorkerCounter.increateAndGet(workerResourceSpec);
```
increate -> increate.

xintongsong commented on flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/ActiveResourceManagerTest.java:
```diff
@@ -0,0 +1,75 @@
+
+		try {
+			counter.decreaseAndGet(spec2);
+			fail("Should have failed calling `decreaseAndGet` on zero.");
```
That might cover problems if another `IllegalStateException` is thrown from previous calls to the `decreaseAndGet`.

KarmaGYZ replied:
Even so, I think we should only catch IllegalStateException instead of Throwable.

xintongsong replied:
In this case, I would be ok for a Throwable of any kind, as long as this call explicitly fails.

KarmaGYZ replied:
I think we could split it into two test cases. One tests the behavior of increase and decrease while the other tests that it will fail in calling `decreaseAndGet` on zero. This avoids mixing multiple tests within a single test.

xintongsong replied:
That sounds good.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ActiveResourceManager.java:
```diff
@@ -114,4 +121,42 @@ public ActiveResourceManager(
+
+		public int increateAndGet(final WorkerResourceSpec workerResourceSpec) {
+			return pendingWorkerNums.compute(
+				Preconditions.checkNotNull(workerResourceSpec),
```
Not sure what is the problem.
The argument is checked before used, and before any changes or calculations this method performs.

KarmaGYZ replied:
The stack trace would be a little different.
```
java.lang.NullPointerException
	at org.apache.flink.util.Preconditions.checkNotNull(Preconditions.java:58)
	at increateAndGet
```
vs
```
java.lang.NullPointerException
	at org.apache.flink.util.Preconditions.checkNotNull(Preconditions.java:58)
	at Test.lambda$increateAndGet$0
	at java.util.HashMap.compute(HashMap.java:1197)
	at increateAndGet
```
I just think there would be two useless lines. I do have a strong opinion though.

xintongsong replied:
I don't think that is a problem.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ActiveResourceManager.java:
```diff
@@ -114,4 +121,42 @@ public ActiveResourceManager(
+
+		public int decreaseAndGet(final WorkerResourceSpec workerResourceSpec) {
+			final Integer newValue = pendingWorkerNums.compute(
+				Preconditions.checkNotNull(workerResourceSpec),
```
Same here.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ActiveResourceManager.java:
```diff
@@ -114,4 +121,42 @@ public ActiveResourceManager(
+					Preconditions.checkState(num > 0);
+					return num == 1 ? null : num - 1;
+				});
+			return newValue != null ? newValue : 0;
```
That is already done by returning `null` in the lambda of `compute`.

xintongsong commented:
@KarmaGYZ, thanks for the review. Comments addressed.

wangyang0918 commented on flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/ActiveResourceManagerTest.java:
```diff
@@ -0,0 +1,73 @@
+public class ActiveResourceManagerTest {
+
+	@Test
+	public void testPendingWorkerCounterIncraseAndDecrease() {
```
Maybe typo here. `Incrase` -> `Increase`

wangyang0918 commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ActiveResourceManager.java:
```diff
@@ -18,6 +18,7 @@
 
```
This commit should be prior to 7d2ded.

xintongsong replied:
It is.
Github messed it up. It shows commits ordered by the committing time (the time in the commit message), instead of the order in the code base.

xintongsong commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -278,7 +288,7 @@ private void removePodIfTerminated(KubernetesPod pod) {
 			final KubernetesWorkerNode kubernetesWorkerNode = workerNodes.remove(new ResourceID(pod.getName()));
 			if (kubernetesWorkerNode != null) {
-				requestKubernetesPodIfRequired();
+				requestKubernetesPodIfRequired(podWorkerResources.get(pod.getName()));
```
True. We should check for the null value here.

xintongsong commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -81,10 +80,8 @@
-	/** The number of pods requested, but not yet granted. */
-	private int numPendingPodRequests = 0;
+	/** Map from pod name to worker resource. */
+	private final Map<String, WorkerResourceSpec> podWorkerResources;
```
1. True. I'll try to clean up when the pod with corresponding name is removed from `workerNodes`. I believe that means we no longer need its resource information.

2. I think it's by design. The original behavior is that RM asks SM for pending slots, and will only request new pod only if there are more pending slots than the slots of the pending TM/pods. In case of JM failover, there should be no pending slots from previous attempt. Therefore, only slots requested in the new attempt is considered. I think the current behavior aligns with the original.

xintongsong commented:
Thanks for the review, @wangyang0918.

Regarding your concern, I think the current behavior does align with the original one. If a TM is terminated exceptionally, whether we request a new pod is not decided by the terminated TM, but decided by whether SlotManager has more pending slots (pending workers after the changes) than that of ResourceManager's pending TMs/pods. In case of JM failover, if the new SM has not requested any slots/workers from RM, then there will be no pending slots/workers, and RM will not request new TM/pod anyway. If the new SM has requested new worker, then there should be corresponding `WorkerResourceSpec` in `podWorkerResources`.

xintongsong commented:
@wangyang0918
I've addressed your comments, except for those I explained in my reply.

xintongsong commented:
Rebased to resolve conflicts.

xintongsong commented:
@tillrohrmann 

I've addressed your comments in #11353 related to the K8s deployment in this PR. Please take another look at your convenience. Thank you.

I would need some more time addressing the Yarn related PRs and test against various Hadoop versions. So I think it might be easier to review this one separately.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -189,15 +180,17 @@ public boolean stopWorker(final KubernetesWorkerNode worker) {
-					numPendingPodRequests--;
+				WorkerResourceSpec workerResourceSpec = podWorkerResources.get(pod.getName());
+				final int pendingNum = getNumPendingWorkersFor(workerResourceSpec);
+				if (pendingNum > 0) {
```
What happens if `pendingNum <= 0`? Shouldn't we return/stop the pod? This is actually out of the scope for this PR but it still looks not right.

xintongsong replied:
I double checked with @wangyang0918  about on what conditions will `onAdded` be called. There are two possible reasons that `onAdded` is called while there's no pending worker request.
- A duplicated `onAdded` message is received. In such cases, `workerNodes` should already contains the pod name. We can check that and simply ignore the duplicated message.
- `onAdded` message for a pod requested from the previous attempt. It is guaranteed that such pods is already added to `workerNodes` in `recoverWorkerNodesFromPreviousAttempts` before the `onAdded` is called, so we can apply the same check and ignore the duplicated pods.

wangyang0918 replied:
Recently i also realize that the `onAdded` for a same `Pod` may be called more that once here, e.g. some network problems cause the pod watcher reconnect, so it will be good to check the existence first. Otherwise, the `numPendingPodRequests` may be decreased wrongly.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -237,57 +230,73 @@ private void recoverWorkerNodesFromPreviousAttempts() throws ResourceManagerExce
+		final KubernetesTaskManagerParameters parameters =
+			createKubernetesTaskManagerParameters(workerResourceSpec);
+
+		final KubernetesPod taskManagerPod =
+			KubernetesTaskManagerFactory.createTaskManagerComponent(parameters);
+		kubeClient.createTaskManagerPod(taskManagerPod);
```
This is also unrelated to this PR. The naming is a bit unfortunate. We have a `createTaskManagerComponent` which creates a `KubernetesPod` and a `createTaskManagerPod` which does not seem to create/return something. I think `createXYZ` should create `XYZ` and return it. Hence, the first method should be called `createKubernetesPod` and the second could be called `startKubernetesPod` or `startTaskManagerPod`.

wangyang0918 replied:
Sorry to jump to here. I want to add some additional explanations here. 

Why i name it to `createTaskManagerPod` instead of `startTaskManagerPod`?
* we usually use `create` in K8s commands(e.g. `kubectl create -f taskmanager.yaml`). The start process will be done internally and automatically in K8s `kubelet`. 

Why do not return a `Pod` for `createTaskManagerPod`?
* Of cause we could return a created `Pod`. However, it is not a final or stable state. Since the K8s will do some updates for the `Pod`, the ip address, the status. So i do not return the `Pod` to avoid it may be used incorrectly. 

For the `KubernetesJobManagerFactory#createJobManagerComponent` and `KubernetesTaskManagerFactory#createTaskManagerComponent`, maybe we could rename them to `KubernetesJobManagerFactory#createJobManagerSpecification` and `KubernetesTaskManagerFactory#createTaskManagerPod`.

tillrohrmann replied:
I think my main problem is that we have two `createXYZ` methods where the one method is creating an object and the other method doing some side effects. Since they are doing different things, they should not have the same prefix. Which one starts with `createXYZ` does not matter but I would like to break up the seemingly symmetry of these methods which does not exist.

Moreover, I think it is a good practice to include the name of the type a method instantiates in the method name: `KubernetesPod create/instantiate/buildKubernetesPod()` or `JobManagerPodSpecification create/instantiate/buildJobManagerPodSpecification`.

wangyang0918 replied:
Thanks for your explanation. From your suggestion guideline, the following two interfaces are already in good shape.
```
// Create the Master components, this can include the Deployment, the ConfigMap(s), and the Service(s).
void createJobManagerComponent(KubernetesJobManagerSpecification kubernetesJMSpec);
CompletableFuture<Void> createTaskManagerPod(KubernetesPod kubernetesPod);
```

And the following two methods could be renamed.
```
public static KubernetesJobManagerSpecification createJobManagerComponent(KubernetesJobManagerParameters kubernetesJobManagerParameters)
public static KubernetesPod createTaskManagerComponent(KubernetesTaskManagerParameters kubernetesTaskManagerParameters)
```
===>
```
public static KubernetesJobManagerSpecification buildKubernetesJobManagerSpecification(KubernetesJobManagerParameters kubernetesJobManagerParameters)
public static KubernetesPod buildTaskManagerKubernetesPod(KubernetesTaskManagerParameters kubernetesTaskManagerParameters)
```

tillrohrmann replied:
Jup, this would be a good improvement.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -237,57 +230,73 @@ private void recoverWorkerNodesFromPreviousAttempts() throws ResourceManagerExce
+	}
+
+	private KubernetesTaskManagerParameters createKubernetesTaskManagerParameters(WorkerResourceSpec workerResourceSpec) {
+		final TaskExecutorProcessSpec taskExecutorProcessSpec =
```
this variable hides `ActiveResourceManager.taskExecutorProcessSpec`. This can lead to subtle bugs. Hence, I would suggest to rename this field into something like `defaultTaskExecutorProcessSpec`.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -237,57 +230,73 @@ private void recoverWorkerNodesFromPreviousAttempts() throws ResourceManagerExce
 
-		while (requiredTaskManagers > numPendingPodRequests) {
-			requestKubernetesPod();
+		while (requiredTaskManagers-- > pendingWorkerNum) {
```
If, then I would not decrease `requiredTaskManager` but increase `pendingWorkerNum` because that is what `requestKubernetesPod` does.

Ideally, the while condition should look like `while (requiredTaskManagers > getNumPendingWorkersFor(workerResourceSpec))` because now we implicitly encode that `requestKubernetesPod` will increase `getNumPendingWorkersFor` by one. I actually think that this should also not perform worse because usually this loop will only be executed once.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -189,15 +180,17 @@ public boolean stopWorker(final KubernetesWorkerNode worker) {
 			for (KubernetesPod pod : pods) {
-				if (numPendingPodRequests > 0) {
-					numPendingPodRequests--;
+				WorkerResourceSpec workerResourceSpec = podWorkerResources.get(pod.getName());
```
Can this be `null`?

xintongsong replied:
If we add a duplication check to filter out pods from previous attempt, then this should not be `null`. I'll add a `checkNotNull`.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -237,57 +230,73 @@ private void recoverWorkerNodesFromPreviousAttempts() throws ResourceManagerExce
 
 	private void removePodIfTerminated(KubernetesPod pod) {
 		if (pod.isTerminated()) {
 			kubeClient.stopPod(pod.getName());
```
Can it happen that we stop a pod w/o running `onAdded` which decrease the super classes `PendingWorkerCounter`? If yes, then the internal state might go out of sync.

xintongsong replied:
Nope. There are two occurrence of `kubeClient.stopPod` and they're both guaranteed to be called after `onAdded`.

tillrohrmann replied:
So this means that `KubernetesResourceManager.onError` will only be called if `onAdded` has been called before? I guess this is also a question for @wangyang0918.

wangyang0918 replied:
I post the `WatchEvent` api in K8s document[1]. From the definition, i do not find the the guarantee that `Error` event is always returned after `Added`. We already have a [ticket](https://issues.apache.org/jira/browse/FLINK-17177) to track how to handle the `onError` event in `KubernetesResourceManager` correctly. I think we could keep the current behavior and try to find a solution in FLINK-17177.

Moerover, i need to dig into the K8s ApiServer implementation and check when the `onError` `WacherEvent` could be returned.


[1]. https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#watchevent-v1-meta

zhengcanbin replied:
> So this means that `KubernetesResourceManager.onError` will only be called if `onAdded` has been called before? I guess this is also a question for @wangyang0918.

There is no guarantee for this, `ERROR` could be thrown in https://github.com/kubernetes/kubernetes/blob/343c1e7636fe5c75cdd378c0b170b26935806de5/staging/src/k8s.io/apimachinery/pkg/watch/streamwatcher.go#L121

xintongsong replied:
It seems I was wrong about `onError` always called after `onAdded`.
Then isn't `numPendingPodRequests` already possibly go out of sync before this PR? Do we observe any cases?

@zhengcanbin Would it be possible that a pod is never successfully started, that we received `onError` but never `onAdded`, `onModified` or `onDeleted`?

zhengcanbin replied:
> Then isn't `numPendingPodRequests` already possibly go out of sync before this PR? Do we observe any cases?

Yes, I noticed this problem when I dived into `KubernetesResourceManager` yesterday, it is possible that we do not receive an `ADD` Event(we met this kind of problem in our K8s production env, but not in Flink so far) even if the Pod has been deployed successfully on the K8s cluster.

> @zhengcanbin Would it be possible that a pod is never successfully started, that we received `onError` but never `onAdded`, `onModified` or `onDeleted`?

It is possible if something goes wrong in the HTTP stream and the K8s server sends an `ERROR`, but I think this should rarely happen.

xintongsong replied:
It seems there are two issues.
1. `numPendingPodRequests`/`PendingWorkerCounter` may go out of sync, if pod is stopped in `onError` before `onAdded` is called.
2. Whether should we remove the pod or not when `onError` is called.

I would suggest the following.
- Try to resolve (1) in this PR, based on the current behavior that pod will be removed when `onError` is called. We can check whether the pod is requested in the current attempt by looking up in `podWorkerResources`, and whether the `onAdded` has been called by looking up in `workerNodes`. If the pod is requested in the current attempt and `onAdded` is not called yet, then we should decrease the pending count.
- Leave (2) to be discussed and resolved in FLINK-17177. We may or may not need to remove the pod when `onError` is called. Moreover, if we remove the pod and decrease the pending count, should we try to request another new pod? To answer these questions, I guess we need to better understand under what circumstances will an `ERROR` be received.

WDYT? @wangyang0918 @zhengcanbin @tillrohrmann

wangyang0918 replied:
I agree with @xintongsong to make sure the `numPendingPodRequests/PendingWorkerCounter` are correctly increased and decreased in this PR. If the `onError` happens before `onAdded`, then we should also decrease the `numPendingPodRequests`.

For how to handle `onError`, let's keep the discussion in FLINK-17177. Since i find the fabric8 kubernetes client has a different implementation with K8s go client. It will never produce a `onError` `WatchEvent` on client side[1]. But i am not familiar with the K8s apiserver about when it will return a `Error` type `WatchEvent`.

[1]. https://github.com/fabric8io/kubernetes-client/blob/1afea0c364a48d9a8745539f84de59630ef6f559/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/WatchHTTPManager.java#L289

tillrohrmann replied:
I think it is a good idea to split these two efforts. 

In fact for (1), I think we need the same checks for `onModified` and `onDeleted` as well. Hence, `removePodIfTerminated` should check whether it is an intended shutdown or not (imagine a manual termination of a pod). In case that it is not, then it needs to decrease the `PendingWorkerCounter` and call `requestKubernetesPodIfRequired`.

It looks as if `removePodIfTerminated` already does this in the case that a pod has been started (`onAdded` has been called) and if the pod fails then. However, it does not do it in case `onAdded` has not been called. But in this case, I think we should request a new pod or fail fatally.

tillrohrmann replied:
I think it could work by using the `podWorkerResources` map to indicate whether a pod needs to be restarted or not. Whenever it contains an entry for a given `ResourceID` we need to restart it. This, of course, means that we need to clear this entry in `stopWorker`.

xintongsong replied:
I'm actually thinking about similar approach. I think we should do two things for all the stopped / terminated workers.
- Check whether `onAdded` is called by looking into `workerNodes`. If it's not called, we should decrease the pending count.
- Try to restart the pod if required. As for indicating whether to restart the worker or not, I'm in favor of relying on `getRequiredResources` which always fetch the up-to-date number of required resources from `SlotManager`, rather than relying on `podWorkerResources`. But ideally these two should be consistent if we properly clean-up `podWorkerResources`.

I'm trying to rebase this PR onto the latest master for conflict resolving. And there's a newly introduced `internalStopPod` which might be a good place for applying the above two actions for all stopped / terminated workers.

tillrohrmann replied:
I'd also be fine with solving this problem in a separate effort. I think the problem has been part of the `KubernetesResourceManager` before. My suggestion for this effort would be to add thorough test coverage. Moreover, it might make sense to see whether there is similar functionality in the `YarnResourceManager` and to think whether it can/should be moved to the `ActiveResourceManager`. I have the feeling that we are re-implementing logic for different `ResourceManager` implementations.

wangyang0918 replied:
I am afraid that the situation is different in YARN. In `YarnResourceManager`, when the `AMRMClient` heartbeats failed, may because YARN RM internal error, or invalid resource request, or network problems, or sth else, we will call `onFatalError` to fail the current jobmanager attempt and start a new one. We do not retry in such case.

Maybe we could have the same logics in `KubernetesResourceManager` if we confirm that the `Error WatchEvent` is sent only when K8s ApiServer internal error, or invalid pod spec. I would give more information in FLINK-17177 after dive into the K8s ApiServer implementation.

Till is right, we need to add thorough test coverage for this effort and try to move the same logics to `ActiveResouceManager`.

zhengcanbin replied:
@xintongsong Sorry for the late reply. I accidentally jumped into this PR a few days ago and currently, I haven't taken a closer look at this PR. But recently I have two concerns:
1. Maybe our handle logic should not rely on the `EVENT` order since there is no guarantee for this, also, the `EVENT` from the `Watcher` stream may not have exactly-once semantics, we may lose some of them or receive the same `EVNET` more than once.
2. We rely heavily on the informer of `Watcher` to track the state of the Pods, I am not entirely sure that this is a good solution, maybe we need an active polling mechanism as a supplement to get the whole view of the Pods. Of course, it needs further discussion and could be resolved in another ticket if necessary, which is out of the scope of this PR.

xintongsong replied:
Thanks all for the feedbacks and good discussion.
@zhengcanbin, I've added more checks to avoid relying on the order and non-duplication of received events.
@tillrohrmann, I've also added more test cases verifying the pending worker counter with out-of-ordered and duplicated events.

tillrohrmann commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesResourceManagerTest.java:
```diff
@@ -337,6 +342,46 @@ public void testGetCpuCoresNumSlots() throws Exception {
+				final PodList terminatedPodList = kubeClient.pods().list();
+				assertEquals(2, terminatedPodList.getItems().size());
+				assertFalse(terminatedPodList.getItems().contains(initialPod1));
+				assertTrue(terminatedPodList.getItems().contains(initialPod2));
```
Why is `initialPod2` contained in the `terminatedPodList`?

tillrohrmann replied:
Should `terminatedPodList` rather be called `activePodList`?

xintongsong replied:
I meant "pod list after the `terminatePod` action". I should choose better words.

xintongsong commented:
@tillrohrmann 
Thanks for the review. I've addressed the comments.

xintongsong commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -320,5 +333,16 @@ private void internalStopPod(String podName) {
+
+		// If the stopped pod is requested in the current attempt (workerResourceSpec is known) and is not yet added,
+		// we need to notify ActiveResourceManager to decrease the pending worker count.
+		if (workerResourceSpec != null && kubernetesWorkerNode == null) {
```
I think we don't want to call `notifyNewWorkerAllocationFailed` and decrease the pending count when a recovered pod is being stopped. The `pendingWorkerCounter` only accounts for the workers requested in the current  attempt.

tillrohrmann replied:
But what we want to do is to restart a recovered pod if it fails and it is needed. Assume that our cluster recovers a pod from a previous attempt. Then we submit a job and the job uses slots of this pod. Now the pod fails. Unlike in the case of a started pod, Flink won't try to restart this pod. I think there should not be a difference between in the failure handling between started and recovered pods.

xintongsong replied:
I think what we need to restart is the workers that are requested by the slot manager, but not yet registered to the slot manager. Because in such cases, the slot manager does not know that these pods are failed and will keep expecting them to register. I believe that is the behavior before.

For a worker that is already registered, in case of failure:
- If there are slots on the failed worker that are already in use, JM will realize that the slots are lost, and send out new slot requests, which will trigger starting of new workers if needed.
- If none of the slots on the failed worker is in use, then we don't loose anything. New workers will be started when new slot requests are received and cannot be satisfied by registered free slots.

For a pod recovered from the previous attempt, if it is not yet registered, slot manager will not be aware of it at all. Thus, slot manager should already requested enough workers for satisfying the slot requests even without this failed pod. Moreover, we have not increased the `pendingWorkerCounter` for the recovered pod and we cannot do that because we don't know its `workerResourceSpec` until it is registered. So we should also not decrease the `pendingWorkerCounter` when the recovered pod is added/failed, otherwise the `pendingWorkerCounter` will go out of sync.

xintongsong replied:
With `requestKubernetesPodIfRequired` now tries to request pods for all the `workerResourceSpec`s, there's not much differences between in failure handling of started/recovered pods. The only difference is whether we decrease the `pendingWorkerCounter` or not.

tillrohrmann replied:
In the current state with calling `requestKubernetesPodIfRequired` it should now work.

I think we are still lacking a bit of test coverage, though. For example if a recovered pod is being used by a job and if the pod now fails, it should be restarted because the `SlotManager` needs the pod.

xintongsong replied:
If a recovered pod is being used by a job, there could be two kinds of situations.
- If the slots are already offered to JM, then JM will find out that the TM/slots are lost, and re-request the slots if needed.
- If the slots are already assigned by SM but not yet offered to JM, RM will discover the TM failure and notify JM about the allocation failure. If needed, JM will re-request the slot. This behavior is covered by `SlotManagerImplTest#testNotifyFailedAllocationWhenTaskManagerTerminated`.

xintongsong commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -239,70 +247,73 @@ private void recoverWorkerNodesFromPreviousAttempts() throws ResourceManagerExce
+							() -> requestKubernetesPodIfRequired(workerResourceSpec),
+							retryInterval);
+					} else {
+						podWorkerResources.put(parameters.getPodName(), workerResourceSpec);
+						final int pendingWorkerNum = notifyNewWorkerRequested(workerResourceSpec);
```
True.
I'll move these two lines to before `kubeClient.createTaskManagerPod` (which is on the main thread), and clean the states if `kubeClient.createTaskManagerPod` is completed exceptionally.
To guarantee the state cleaning happens also on the main thread and before the retry, I'll wrap it with another `runAsync`.

xintongsong commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -311,7 +322,9 @@ protected double getCpuCores(Configuration configuration) {
 	}
 
-	private void internalStopPod(String podName) {
+	private Optional<WorkerResourceSpec> internalStopPod(String podName) {
```
Could you explain what behaviors do you suggest to verify for this method?
I think this is a private internal method. The code paths invoking this method should be already covered by `testStartAndStopWorker` and `testTaskManagerPodTerminated`.

tillrohrmann replied:
I don't mean to directly test this method but the behaviour in case of pod terminations/failures. For example, I think at the moment we treat the failure of a recovered and a started pod differently. In the former case we won't do anything and in the latter case we will try to restart it. I'm not sure whether this needs to be the case. See https://github.com/apache/flink/pull/11323#discussion_r413767396.

xintongsong commented:
Thanks @tillrohrmann, I've addressed your comments, except for two that I have some questions about.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -251,29 +250,32 @@ private void requestKubernetesPod(WorkerResourceSpec workerResourceSpec) {
 		kubeClient.createTaskManagerPod(taskManagerPod)
 			.whenComplete(
-				(ignore, throwable) -> {
+				(ignore, throwable) -> runAsync(() -> {
```
```suggestion
				(ignore, throwable) -> {
```

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -251,29 +250,32 @@ private void requestKubernetesPod(WorkerResourceSpec workerResourceSpec) {
 		final KubernetesPod taskManagerPod =
 			KubernetesTaskManagerFactory.buildTaskManagerKubernetesPod(parameters);
 		kubeClient.createTaskManagerPod(taskManagerPod)
 			.whenComplete(
```
```suggestion
			.whenCompleteAsync(
```

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -251,29 +250,32 @@ private void requestKubernetesPod(WorkerResourceSpec workerResourceSpec) {
 						log.info("TaskManager {} will be started with {}.", parameters.getPodName(), workerResourceSpec);
 					}
-				}
+				})
```
```suggestion
				},
				getMainThreadExecutor())
```

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java:
```diff
@@ -251,29 +250,32 @@ private void requestKubernetesPod(WorkerResourceSpec workerResourceSpec) {
-							() -> requestKubernetesPodIfRequired(workerResourceSpec),
+							this::requestKubernetesPodIfRequired,
 							retryInterval);
 					} else {
```
I think in order to properly guard this change we need a test where we request a pod, the kube client takes some time to complete the returned future and in the mean time we call `onAdded`. Before, we did not test this and did not catch the problem.

xintongsong commented:
Thanks for the review, @tillrohrmann.
I've added another fixup commit addressing your comments. The PR has been rebased onto the recent master branch for conflict resolving.

tillrohrmann commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesResourceManagerTest.java:
```diff
@@ -634,4 +683,15 @@ private FlinkKubeClient createTestingFlinkKubeClientAllocatingPodsAfter(
+		return new Fabric8FlinkKubeClient(flinkConfig, kubeClient, () -> kubeClientExecutorService) {
+			@Override
+			public CompletableFuture<Void> createTaskManagerPod(KubernetesPod kubernetesPod) {
+				return super.createTaskManagerPod(kubernetesPod).runAfterBoth(trigger, () -> {});
```
Nice :-)
